src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -933,6 +985,58 @@ public DocumentId AddTextContainer(SourceTextContainer textContainer, string ful
                 return documentId;
             }
 
+            public Task AddSourceFileInfoAsync(string filePath, SourceFileInfo fileInfo, ImmutableArray<string> folders)
+            {
+                var documentInfo = CreateDocumentInfoFromFileInfo(fileInfo, folders.IsDefault ? null : (IEnumerable<string>)folders);
+                var documentId = documentInfo.Id;
+
+                lock (_project._gate)
+                {
+                    if (_documentPathsToDocumentIds.ContainsKey(filePath))
+                    {
+                        throw new ArgumentException($"'{filePath}' has already been added to this project.", nameof(filePath));
+                    }
+
+                    _documentPathsToDocumentIds.Add(filePath, documentId);
+
+                    _project._fileWatchingTokens.Add(documentId, _project._documentFileChangeContext.EnqueueWatchingFile(fileInfo));
+
+                    if (_project._activeBatchScopes > 0)
+                    {
+                        _documentsAddedInBatch.Add(documentInfo);
+                    }
+                    else
+                    {
+                        _project._workspace.ApplyChangeToWorkspace(w => _documentAddAction(w, documentInfo));
+
+                        // for now, all document added this way is considered closed and never can be opened.
+                        // we will figure on clsoe/open transition in next round when we figure out how to support
+                        // modifications on generated files.
+                        //
+                        // _project._workspace.CheckForOpenDocuments(ImmutableArray.Create(filePath));
+                    }
+                }
+
+                return Task.CompletedTask;
+            }
+
+            public Task RemoveNonSourceFileAsync(string filePath, CancellationToken cancellationToken)
+            {
+                lock (_project._gate)
+                {
+                    if (!_nonSourceFilePathToSourceFileInfo.TryGetValue(filePath, out var fileInfo))
```

Comment:
I don't think there's any code that adds to `_nonSourceFilePathToSourceFileInfo` so the code that processes changes will never see a change.

Reply:
:) I just realized as well :) sorry about that.